Sources/NIO/FileHandle.swift

```diff
@@ -81,12 +81,56 @@ public final class NIOFileHandle: FileDescriptor {
 }
 
 extension NIOFileHandle {
+    /// `Mode` represents file access modes.
+    public struct Mode: OptionSet {
+        public let rawValue: UInt8
+
+        public init(rawValue: UInt8) {
+            self.rawValue = rawValue
+        }
+
+        internal var posixMode: CInt {
+            switch self {
+            case [.read, .write]:
+                return O_RDWR
+            case .read:
+                return O_RDONLY
+            case .write:
+                return O_WRONLY
+            default:
+                preconditionFailure("Unsupported mode value")
+            }
+        }
+
+        /// Opens file for reading
+        public static let read = Mode(rawValue: 1 << 0)
+        /// Opens file for writing
+        public static let write = Mode(rawValue: 1 << 1)
+    }
+
+    public struct Flags {
+        internal var posixMode: mode_t
+        internal var posixFlags: CInt
+
+        public static let `default` = Flags(posixMode: 0, posixFlags: 0)
+
+        /// Allows file creation when opening file for writing. File owner is set to the effective user ID of the process.
+        ///
+        /// - parameters:
+        ///     - posixMode: `file mode` applied when file is created. Default permissions are: read and write for file owner, read for owners group and others.
+        public static func allowFileCreation(posixMode: mode_t = S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH) -> Flags {
+            return Flags(posixMode: posixMode, posixFlags: O_CREAT)
+        }
+    }
+
     /// Open a new `NIOFileHandle`.
     ///
     /// - parameters:
-    ///     - path: the path of the file to open. The ownership of the file descriptor is transferred to this `NIOFileHandle` and so it will be closed once `close` is called.
-    public convenience init(path: String) throws {
-        let fd = try Posix.open(file: path, oFlag: O_RDONLY | O_CLOEXEC)
+    ///     - path: The path of the file to open. The ownership of the file descriptor is transferred to this `NIOFileHandle` and so it will be closed once `close` is called.
+    ///     - mode: Access mode. Default mode is `.readOnly`.
+    ///     - flags: Additional POSIX flags.
+    public convenience init(path: String, mode: Mode = .read, flags: Flags = .default) throws {
+        let fd = try Posix.open(file: path, oFlag: mode.posixMode | O_CLOEXEC | flags.posixFlags, mode: flags.posixMode)
```

Comment:
This looks a bit weird, which makes me realise that `posixMode` is overloaded: there is `Mode.posixMode` and `Flags.posixMode`, but they don't go to the same place. Can we re-evaluate those names? Maybe `Mode.posixMode` should actually be `mode.posixFlags` in order to match `Flags`?

Reply:
Fixed!

Reply:
Wow I always forget how annoying a protocol `OptionSet` is.